crates/staged-sync/src/builder.rs

```diff
@@ -0,0 +1,317 @@
+    stages::{
+        bodies::BodyStage, execution::ExecutionStage, hashing_account::AccountHashingStage,
+        hashing_storage::StorageHashingStage, headers::HeaderStage, merkle::MerkleStage,
+        sender_recovery::SenderRecoveryStage, total_difficulty::TotalDifficultyStage,
+    },
+    Pipeline, PipelineEvent,
+};
+
+use std::sync::Arc;
+use tokio::sync::mpsc::Sender;
+
+use eyre::Result;
+
+/// Rough API usage:
+///
+/// 1. To launch the pipeline
+/// RethBuilder::new(..).with_execution(..).online(..).with_headers(..).with_bodies(..).build()
+/// RethBuilder::new(..).with_execution(..).build()
+///
+/// OR
+///
+/// 2. To make it easy to run an individual stage:
+///
+/// let builder = RethBuilder::new(..).online(..).with_headers(..);
+/// let header = builder.headers_stage().await?.expect("should be configured");
+/// let tx = Transaction::new(builder.db());
+/// let input = ExecInput {
+///     previous_stage: Some((StageId("No Previous Stage"), self.to)),
+///     stage_progress: Some(self.from),
+/// };
+/// stage.execute(&mut tx, input).await?;
+#[must_use = "need to call `build` on this struct"]
+#[derive(Debug, Default)]
+pub struct RethBuilder {
+    senders_recovery: Option<SenderRecoveryConfig>,
+
+    chain_spec: Option<ChainSpec>,
+    execution: Option<ExecutionConfig>,
+
+    merklize: bool,
+
+    channel: Option<Sender<PipelineEvent>>,
+}
+
+impl RethBuilder {
+    /// Instantiates the builder with the [`Default`] implementation.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Converts the RethBuilder to an online one to be integrated with the Headers/Bodies stages
+    pub fn online<C, N>(self, consensus: C, network: N) -> OnlineRethBuilder<C, N> {
+        OnlineRethBuilder::new(self, consensus, network)
+    }
+
+    /// Configures the [`SenderRecoveryStage`]
+    pub fn with_senders_recovery(mut self, config: SenderRecoveryConfig) -> Self {
+        self.senders_recovery = Some(config);
+        self
+    }
+
+    /// Retrieves the [`SenderRecoveryStage`] if it was configured
+    pub fn senders_recovery(&self) -> Option<SenderRecoveryStage> {
+        self.senders_recovery.as_ref().map(|config| SenderRecoveryStage {
+            batch_size: config.batch_size,
+            commit_threshold: config.commit_threshold,
+        })
+    }
+
+    /// Configures the [`ExecutionStage`] parameters. The Default chainspec is mainnet,
+    /// use [`Self::with_chain_spec`] to configure for other chains.
+    pub fn with_execution(mut self, config: ExecutionConfig) -> Self {
+        self.execution = Some(config);
+        self
+    }
+
+    /// Configures [`ExecutionStage`]'s [`ChainSpec`]
+    pub fn with_chain_spec(mut self, chain_spec: ChainSpec) -> Self {
+        self.chain_spec = Some(chain_spec);
+        self
+    }
+
+    /// Set a channel the pipeline will transmit events over (see [PipelineEvent]).
+    pub fn with_channel(mut self, sender: Sender<PipelineEvent>) -> Self {
```

Comment:
I think it would be better to rename this to `event_channel` or `event_listener`